splice_spark2/src/main/spark2.4/com/splicemachine/spark2/splicemachine/SplicemachineContext.scala

```diff
@@ -0,0 +1,832 @@
+    insert(rdd, schema, schemaTableName, Map("insertMode"->"INSERT","statusDirectory"->statusDirectory,"badRecordsAllowed"->badRecordsAllowed.toString) )
+
+  /**
+    *
+    * Insert a dataFrame into a table (schema.table).  This corresponds to an
+    *
+    * insert into from select statement
+    *
+    * @param dataFrame input data
+    * @param schemaTableName output table
+    */
+  def insert(dataFrame: DataFrame, schemaTableName: String): Unit = insert(dataFrame.rdd, dataFrame.schema, schemaTableName)
+
+  /**
+   * Insert a RDD into a table (schema.table).  The schema is required since RDD's do not have schema.
+   *
+   * @param rdd input data
+   * @param schema
+   * @param schemaTableName
+   */
+  def insert(rdd: JavaRDD[Row], schema: StructType, schemaTableName: String): Unit = insert(rdd, schema, schemaTableName, Map[String,String]())
+
+  private[this] def columnList(schema: StructType): String = SpliceJDBCUtil.listColumns(schema.fieldNames)
+  private[this] def schemaString(schema: StructType): String = SpliceJDBCUtil.schemaWithoutNullableString(schema, url).replace("\"","")
+
+  private[this] def insert(rdd: JavaRDD[Row], schema: StructType, schemaTableName: String, spliceProperties: scala.collection.immutable.Map[String,String]): Unit = {
+    val topicName = getRandomName() //Kafka topic
+
+//    println( s"SMC.insert topic $topicName" )
+
+    // hbase user has read/write permission on the topic
+
+    sendData(topicName, rdd, schema)
+
+    val colList = columnList(schema)
+    val sProps = spliceProperties.map({case (k,v) => k+"="+v}).fold("--splice-properties useSpark=true")(_+", "+_)
+    val sqlText = "insert into " + schemaTableName + " (" + colList + ") "+sProps+"\nselect " + colList + " from " +
+      "new com.splicemachine.derby.vti.KafkaVTI('"+topicName+"') " +
+      "as SpliceDatasetVTI (" + schemaString(schema) + ")"
+
+//    println( s"SMC.insert sql $sqlText" )
+    executeUpdate(sqlText)
+  }
+
+  private[this] def sendData(topicName: String, rdd: JavaRDD[Row], schema: StructType): Unit =
+    rdd.rdd.mapPartitions(
+      itrRow => {
+        //println( s"SMC.sendData $topicName" )
+        //println( s"SMC.sendData mapPartitions" )
+        val props = new Properties
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaServers)
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "spark-producer-"+java.util.UUID.randomUUID() )
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[IntegerSerializer].getName)
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[ExternalizableSerializer].getName)
+
+        val producer = new KafkaProducer[Integer, Externalizable](props)
+
+        var count = 0
+        itrRow.foreach( row => {
+          //println( s"SMC.sendData record $count" )
+          producer.send( new ProducerRecord(topicName, count, externalizable(row, schema)) )
```

Comment:
This is not handling task retries

Reply:
Fixed in new commit [0bd8e29](https://github.com/splicemachine/spliceengine/pull/3391/commits/0bd8e2964e1366bb644d9ef9faae11602bf00af5).